intro/README.md

```diff
@@ -0,0 +1,561 @@
+
+2. Overloading:
+
+    ```d
+    void fun(int a) {}
+    void fun(int a, int b) {}
+    ```
+    
+    Overloading is done based on the **type of parameters**, not on the **return type**.
+    
+3. Default parameters:
+
+    ```d
+    void fun(int a, int b=8) {}
+     
+    void main()
+    {
+        fun(7);    // calls fun(7, 8)
+        fun(2, 3); // calls fun(2, 3) 
+    }
+    ```
+        
+    A function may have any number of default parameters. If a default parameter is given, all following parameters must also have default values.
+    
+4. Auto return type:
+
+    ```d
+    auto fun()
+    {
+        return 7;
+    }
+     
+    void main()
+    {
+        int b = fun();
+        auto c = fun();   // works for variables also
+    }
+    ```
+        
+    Auto functions have their return type inferred based on the type of the return statements.
+    Auto can be also used to infer the type of a variable declaration.
+    
+Navigate to **demo/voldemort** directory. Inspect and run the code inside **voldermort.d**.
+
+Now navigate to **intro/voldemort** directory. 
+**Result** is a struct declared inside **fun's** scope.
+However we want to be able to use it in **main** as well.
+Inspect the file. Does it compile? Why? Fix the issue.
+    
+## Unittests
+
+Unit tests are a builtin framework of test cases applied to a module to determine if it is working properly.
+A D program can be run with unit tests enabled or disabled.
+
+Unit tests are a special function defined like:
+
+```d
+unittest
+{
+    // test code
+}
+```
+
+Individual tests are specified in the unit test using [assert expressions](https://dlang.org/spec/expression.html#AssertExpression).
+There can be any number of unit test functions in a module, including within struct and class declarations.
+They are executed in lexical order.
+Unit tests, when enabled, are run after all static initialization is complete and before the **main()** function is called.
+
+```d
+struct Sum
+{
+    int add(int x, int y) { return x + y; }
+ 
+    unittest
+    {
+        Sum sum;
+        assert(sum.add(3,4) == 7);
+        assert(sum.add(-2,0) == -2);
+    }
+}
+```
+
+Navigate to the **demo/binary-search** directory.
+Inspect the source file binarySearch.d.
+As the name implies, a binarySearch algorithm is implemented on integers.
+Compile and run the file.
+
+- Write a unittest function that tests some corner cases. Are there any bugs in the algorithm implementation? If yes, fix them.
+- Rewrite the binarySearch algorithm to make use of slices.
+   
+## Contract Programming
+
+Contracts are a breakthrough technique to reduce the programming effort for large projects. Contracts are the concept of preconditions, postconditions and errors.
+
+### Preconditions/Postconditions
+
+The pre contracts specify the preconditions before a statement is executed. The most typical use of this would be in validating the parameters to a function.
```

Comment:
```suggestion
The pre contracts specify the preconditions before a statement is executed.
The most typical use of this would be in validating the parameters to a function.
```